### DISCUSSION TITLE
Roadmap to V1

### DISCUSSION THREAD

**Discussion by aldoborrero** [nix-community/ethereum.nix#326] [General]
## Current situation

Since the project's inception, we've devoted our spare time to the development of Ethereum.nix in a somewhat informal manner. It began with my own efforts to supplement missing packages not found in `nixpkgs` with the intention of upstreaming them. Shortly thereafter, Brian joined the efforts and began improving the modules. Sergey followed suit with consistent contributions.

Now, I envision a future where Ethereum.nix becomes an integral tool for:

- Spinning up a custom dev environment to quickly test features on a bootstrapped PoS / PoW with major client support and peripheral tooling (i,e: `mev-boost` and friends).
- Efficient development of Solidity / Vyper contracts with instant access to virtually all tooling, including auditing tools like [Slither](https://github.com/crytic/slither), [Mythril](https://github.com/ConsenSys/mythril), [Brownie](https://github.com/eth-brownie/brownie), and [ApeWorx](https://www.apeworx.io/). Also, for security reasons, to support not only previous versions of Solidity / Vyper compilers (sometimes you don't want to target the latest).
- Effortlessly incorporate any of these tools into any CI pipeline. By developing common CI actions or alternatives, we eliminate the need for Docker, streamlining workflows ([these actions](https://github.com/crytic/slither-action) can [be rewritten quite easily to use Ethereum.nix and direnv](https://github.com/aldoborrero/use-nix-action)).
- Deploying a PoS staking platform using our custom modules for a secure default setup on different cloud / bare-metal providers.

Ethereum.nix aims to become the go-to tool for all things Ethereum.

Recently, we've [considered the potential of donations and grants](https://github.com/nix-community/ethereum.nix/actions) to fortify the sustainability of the project. Such an approach could buy us precious time, our scarcest resource, allowing us to concentrate on new feature development, bug fixes, stability enhancements, issue triage, and general support.

Before taking that step, however, I firmly believe we need to refine certain aspects of the project. This will make Ethereum.nix more user-friendly, especially for those less technically versed with nix.

## Next steps

In order to apply for any grants, the following should happen:

- [x] Let's revamp our `Ethereum.nix Planning board` and establish a milestone specifically for V1. Instead of solely depending on GitHub issues to track our project, I suggest we make more frequent use of this tool. To streamline the process, I removed our custom labels associated with effort estimation and similar metrics a few weeks ago, considering them redundant when using the planning board.
- [x] Prioritizing top-tier documentation is crucial, and we must dedicate significant effort to ensure its quality. Personally, it's on what I'm focused on these days (I'll flesh out the tasks again to include them on the V1 milestone):
  - I've revamped the design aesthetics of our documentation (see Screenshot 1).
  - I'm reworking the documentation guides to make it easier to get started (especially if you're new to nix).
  - I'm developing a custom `mkdocs` plugin to streamline the extraction of `nixosModules` options with a much better design than the current solution (see Screenshot 2 vs Screenshot 3).
  - I've introduced better templates for opening Github tickets with clear distinctions for improving documentation, adding support to a new tool, or reporting bugs.
- [ ] Ideally, we should add support to, at least, some more components, by introducing `nixosModules` for at least `Vouch` and `Dirk` and also trying to include packaging `nimbus-eth`.

I welcome your insights regarding any overlooked considerations or necessary actions. Your thoughts and opinions on the concept are also greatly appreciated.

### Screenshots 

**Screenshot 1**:

![image](https://github.com/nix-community/ethereum.nix/assets/82811/d01d2f43-098e-48a3-8304-0aa289a1662b)

**Screenshot 2**:

![image](https://github.com/nix-community/ethereum.nix/assets/82811/6d3850b3-c03e-455d-84b5-0f86374a121e)

**Screenshot 3**

![image](https://github.com/nix-community/ethereum.nix/assets/82811/b024ea18-b090-4c66-9b6e-ceb9da9083f7)

**Comment by aldoborrero** [comment]
I would like to know your feedback @selfuryon @brianmcgee

**Comment by brianmcgee** [comment]
I largely agree with what you've said and the approach you've outlined. I like the docs improvements and the roadmap makes sense. 

My only concern I guess is the scope. Being able to fund this work more directly will help us to assign dedicated time, but speaking for myself and my other commitments, I think we should to try and find a few more people interested in helping to maintain the project. 

This will not only help with resourcing, but help address the fact that we're looking to take on quite a wide scope. Being able to subdivide responsibilities across a wider group of people will help ensure each area is getting dedicated focus.

**Comment by selfuryon** [comment]
Docs UI improvements are significant. That's cool! The roadmap is good too. I can focus on vouch/dirk modules on this week because I use them in my current work, so I need them anyway.

I also had a conversation with my current company, and they are ready to sponsor a module enhancement in the near future, so maybe it will be our first grant :)

**Reply by aldoborrero** [reply]
@selfuryon I have marked #90 and #89 as in progress on the board. 

In terms of the sponsorship that's great. I have also been in talks with other people that might help us as well.
